src/core/qgsuserprofilemanager.h

```diff
@@ -0,0 +1,172 @@
+     */
+    QgsUserProfileManager( const QString &rootLocation = QString() );
+
+    /**
+     * Resolves the profiles folder for the given path. All paths will have \\profiles appended to the path
+     * Checks QGIS_CUSTOM_CONFIG_PATH first.
+     * @param basePath The base path to resolve the path from.  If empty will use QtStandardPaths to find OS
+     * app data location.
+     * @return The root path to store user profiles.
+     */
+    static QString resolveProfilesFolder( const QString  &basePath = QString() );
+
+    /**
+     * Return the profile from the given root profile location.
+     * If no name is given it returns a profile called "default".
+     * By default will create the profile folder if not found.
+     * By default will init the user settings.
+     * @param defaultProfile The profile name to find. Empty profile name will return "default" for the name.
+     * @param createNew Create the profile folder if it doesn't exist.
+     * @return The user profile
+     */
+    QgsUserProfile *getProfile( const QString &defaultProfile = "default", bool createNew = true, bool initSettings = true );
+
+    /**
+     * Set the root profile location for the profile manager. All profiles are loaded from this
+     * location. Will also contain a profiles.ini for holding profile settings.
+     * @param rootProfileLocation Path to the top level profile folder which contains folders for each profile.
+     */
+    void setRootLocation( QString rootProfileLocation );
+
+    /**
+     * Returns the path to the root profiles location.
+     * @return The root path to the profiles folder.
+     */
+    QString rootLocation() { return mRootProfilePath; }
+
+    /**
+     * Check if the root location has been set for the manager.
+     * @return True if the root location has been set.
+     */
+    bool rootLocationIsSet() const;
+
+    /**
+     * A list of all found profile names.
+     * @return
+     */
+    QStringList allProfiles() const;
+
+    /**
+     * Check if a profile exists.
+     * @return False if the profile can't be found.
+     */
+    bool profileExists( const QString &name ) const;
+
+    /**
+     * Returns the name of the default profile that has been set in .default.
+     * @return The name of the default profile.
+     */
+    QString defaultProfileName() const;
+
+    /**
+     * Sets the default profile name.  The default profile name is used when loading QGIS
+     * with no arguments.
+     * @param name The name of the profile to save.
+     */
+    void setDefaultProfileName( const QString &name );
+
+    /**
+     * Set the default profile name from the current active profile.
+     */
+    void setDefaultFromActive();
+
+    /**
+     * Return the profile found for a given name.
+     * @param name The name of the profile to return.
+     * @return A QgsUserprofile pointing to the location of the user profile.
+     */
+    QgsUserProfile *profileForName( const QString name ) const;
```

Comment:
needs SIP_FACTORY annotation (and better note in the docs that it returns a new profile, caller responsible for cleanup)